### PR TITLE
BGDIINF_SB-2380: Fixed collision and short ID length #major

### PR DESCRIPTION
**Pull request by ltshb**
The sort ID was initially (when produced by mf-chsdi3 project) using the
%x formatting (hexadecimal). This formatting has been forgotten during the
migration of the service into this micro-service repository. This had a negative
impact on the size of the short ID (size increased from 10 to 12 characters).

Also the implementation did not handle short id collisions as it was on mf-chsdi3.
On dynamodb the simplest collision setup is to use the primary key uniqueness
check (other attributes cannot be checked as unique). On the old implementation
on mf-chsdi3 this check was not done which resulted in overwrite in case of
collision. In the new service, because the primary key was not the shortlink_id
a collision resulted with duplicate shortlink_id in the table !

Now we use the shortlink_id as primary key and we make sure that collision
won't overwrite, but trigger retries instead. We allow up to 10 retry before
failing.

Also cleaned up the code with a lots of simplification.

:warning: This is a `#major` breaking changes as it requires a new DynamoDB table
having the shortlink_id attribute set as primary key.

Also updated all dependencies and python to 3.9

### CODE REVIEW

File: tests/unit_tests/test_helpers.py [hunk 1]
```diff
@@ -1,91 +1,158 @@
...
+        # third call
+        mock_generate_short_id.side_effect = ['2', '2', '3']
+        url1 = 'https://www.example/test-one-duplicate-id-first-url'
+        url2 = 'https://www.example/test-one-duplicate-id-second-url'
+        entry1 = self.db.add_url_to_table(url1)
+        self.assertEqual(entry1['shortlink_id'], '2')
+        entry2 = self.db.add_url_to_table(url2)
+        self.assertEqual(entry2['shortlink_id'], '3')
+
+    # @params(1, 2)
```
[the remark this answers is not in the document]
Comment (ltshb) [reply tests/unit_tests/test_helpers.py:120]: @hansmannj I used this code to test the collision with `nanoid` package, once with an ID of length 1 and then of length 2. I thought I leave it here for the moment. I'm planning to write a small ADR regarding the short ID algorithm and I can then move this into the ADR.

File: app/helpers/utils.py [hunk 2]
```diff
@@ -43,5 +59,58 @@ def get_registered_method(app, url_rule):
...
+        code,
+    )
+    return response
+
+
+def get_url():
+    """
+    Get and check the url parameter
+
+    Abort with a 400 status code if there is no url, given to shorten
```
[the remark this answers is not in the document]
Comment (ltshb) [reply app/helpers/utils.py:91]: I don't like too much documenting as comment tends not to be updated, IMHO the name of the functions is already self documenting.

[the remark this answers is not in the document]
Comment (hansmannj) [reply app/helpers/utils.py:91]: Just meant as part of the normal docstring, which contains the return, too.

**Review by hansmannj — APPROVED** [review]
Thanks for fixing and cleaning up! :1st_place_medal:

**Pull request merged**